File: apps/desktop/src/lib/entry-metrics.test.ts
import { describe, expect, it } from 'vitest';
import { countMixedWords, rowWordCount } from './entry-metrics';

describe('countMixedWords', () => {
  it('counts CJK characters one by one', () => {
    expect(countMixedWords('你好世界')).toBe(4);
  });

  it('counts a Latin run as ONE word, not one per letter', () => {
    expect(countMixedWords('hello world')).toBe(2);
  });

  it('mixed CJK + Latin: each CJK char counts, each Latin run counts once', () => {
    // "今天 hello 天气" → 今 天 [hello] 天 气 = 5
    expect(countMixedWords('今天 hello 天气')).toBe(5);
  });

  it('empty / whitespace-only / null / undefined all count as 0', () => {
    expect(countMixedWords('')).toBe(0);
    expect(countMixedWords('   \n\t  ')).toBe(0);
    expect(countMixedWords(null)).toBe(0);
    expect(countMixedWords(undefined)).toBe(0);
  });

  it('numbers count; a punctuation-only run does not (CHANGED 2026-08-14: was 4)', () => {
    // 🔴 In-place correction (原地更正) 2026-08-14. Original assertion and its
    // reasoning, kept verbatim
    // because it recorded the behaviour of the day it was written:
    //   > "v0.2.36," is one run (1), 好 and 的 are one CJK char each (2), and the
    //   > trailing "!" — NOT absorbed into the preceding CJK char, because CJK chars
    //   > are consumed one at a time and only LATIN runs merge with each other — is
    //   > its own 1-char run (1). Total 4.
    // Every clause of that is still true EXCEPT the last: the lone "!" run is
    // still its own run, but a run with no letter and no digit now scores 0
    // rather than 1. "v0.2.36," keeps its 1 (it has letters and digits), so the
    // total moves 4 → 3. The new number is the right one: this row shows the
    // user "3 characters" (3 字) for three countable things — one token, two characters (字) — where it used
    // to count the exclamation mark as a fourth.
    expect(countMixedWords('v0.2.36, 好的!')).toBe(3);
  });

  it('a trailing punctuation mark adds nothing (CHANGED 2026-08-14: was 2)', () => {
    // 🔴 In-place correction (原地更正) 2026-08-14. This test used to be titled "documents the boundary
    // case: CJK punctuation splits an otherwise-adjacent Latin run" and asserted
    // 2, with this reasoning, kept verbatim:
    //   > "好的!" — 好/的 are CJK (2), "!" is latin-ish and adjacent to 的 with no CJK
    //   > between it and the previous char, so it is absorbed into a run starting at
    //   > "!" — this is the "not linguistically correct" corner case the header
    //   > comment calls out.
    // The splitting behaviour it documents is unchanged and still deliberate;
    // what changed is that the resulting punctuation-only run scores 0. So the
    // corner case the header called out is now half-closed: punctuation still
    // BREAKS a run, but it no longer INVENTS a word.
    expect(countMixedWords('好!')).toBe(1); // 好=1, "!"=0
  });

  it('🔴 a run with no letter and no digit counts 0 — it is not a word', () => {
    // The defect this closes: standalone CJK punctuation is not in the CJK class
    // (that class is ideographs and kana only), so it fell into the non-CJK-run
    // branch and scored 1 apiece. A row reading "！！！" reported "1 character".
    expect(countMixedWords('！！！')).toBe(0);
    expect(countMixedWords('，')).toBe(0);
    expect(countMixedWords('，。！？')).toBe(0);
    expect(countMixedWords('... ')).toBe(0);
    expect(countMixedWords('---')).toBe(0);
    // …and the same rule inside real text: four 字, two marks, four counted.
    expect(countMixedWords('你好。世界！')).toBe(4);
  });

  it('punctuation ABSORBED into a lettered run still scores that run once', () => {
    // The other half of the same rule, and the reason the guard tests the whole
    // RUN rather than each character: 「，world」 is a single run (the fullwidth
    // comma is not CJK and there is no space to break it), it contains letters,
    // so it scores 1. Total 3 — the same number as before this change, but now
    // for a reason that survives the comma being moved.
    expect(countMixedWords('你好，world')).toBe(3);
    // Move the comma off the word and the difference shows: it scores 0 alone,
    // where before 2026-08-14 this string counted 4.
    expect(countMixedWords('你好， world')).toBe(3);
  });

  it('🔴 non-ASCII Latin: accented letters count, they are not punctuation', () => {
    // Pinned because it is exactly what the ASCII-only mobile rule gets wrong
    // (apps/mobile/.../entry_metrics.dart counts only [0-9A-Za-z], so it reads
    // "Grüße" as two runs). `\p{L}` is what makes de/fr/es correct here.
    expect(countMixedWords('Grüße aus München')).toBe(3);
    expect(countMixedWords('¿Cómo estás?')).toBe(2);
  });

  it('🔴 Cyrillic counts as words, not as zero', () => {
    // Same reason, harder failure: under an ASCII-only letter class an entire
    // Russian utterance counts 0 — the row would say "0 characters" about text plainly
    // visible above it. ru shipped as a UI language 2026-08-14.
    expect(countMixedWords('привет мир')).toBe(2);
    expect(countMixedWords('Привет, мир!')).toBe(2);
  });

  it('a lone Japanese sentence counts one per glyph, same as Chinese', () => {
    expect(countMixedWords('こんにちは')).toBe(5);
  });

  it('🔴 Korean counts 어절 (space-delimited), NOT syllables — CHANGED 2026-08-14', () => {
    // Was 7 / 5. Modern Korean is written with spaces between words (띄어쓰기),
    // which is precisely what separates it from Chinese and Japanese; the unit
    // has a name — 어절 (eojeol), the spacing unit — and the translation
    // industry counts that. Counting Hangul syllables is the Korean equivalent
    // of counting English letters. Hangul syllables were therefore removed from
    // the `CJK` class. 🔴 This is a judgement call made under owner's
    // 2026-08-14 authorisation to research and correct counting conventions —
    // NOT a ruling owner issued, and it moves a number the user sees. Reversal
    // is one line; see §C of the file header.
    expect(countMixedWords('한국어 단어 세기')).toBe(3); // was 7
    expect(countMixedWords('안녕하세요')).toBe(1); // was 5
    // Korean mixed with Han keeps both rules: the Hangul run is one 어절, the
    // two ideographs are one 字 each.
    expect(countMixedWords('한자 漢字')).toBe(3);
  });

  it('the CJK class itself is unchanged — ideographs and kana still count per char', () => {
    // Boundary guard. Hangul left the class on 2026-08-14; nothing else did, and
    // this pins the ranges so a future edit cannot quietly take another block
    // with it. The compatibility-ideograph pair is written as backslash-u
    // escapes on purpose, and is deliberately not spelled out as literals in
    // this comment either: retyping U+F900 through anything that applies
    // Unicode NFC rewrites it to U+8C48, which moves the range while the line
    // still looks right. Measured hazard, 2026-08-14.
    expect(countMixedWords('㐀䶿')).toBe(2); // Extension A ends
    expect(countMixedWords('一鿿')).toBe(2); // Unified ideographs ends
    expect(countMixedWords('\uF900\uFADF')).toBe(2); // compatibility ideographs
    expect(countMixedWords('぀ヿ')).toBe(2); // kana ends
    expect(countMixedWords('ㇰㇿ')).toBe(2); // katakana phonetic ext
  });

  it('an emoji-only utterance counts 0 — knowingly accepted', () => {
    // A consequence of the letter-or-digit guard, written down rather than left
    // to surprise someone: an emoji is neither a character (字) nor a word (词), and 0 is the
    // honest answer for a length estimate whose unit is one of those two. The
    // row still renders its text; only the count says 0. Before 2026-08-14 this
    // scored 1. If a product decision ever wants emoji counted, this is the
    // test that has to change, and it should change deliberately.
    expect(countMixedWords('😀😀')).toBe(0);
    // …but an astral LETTER (Han Extension B, one surrogate pair) still counts:
    // the guard tests the run with the `u` flag, so it sees one \p{L}, not two
    // lone surrogates.
    expect(countMixedWords('\u{20BB7}')).toBe(1);
  });
});

describe('rowWordCount — the owner "each mode\'s final result" (各模式最终结果) criterion', () => {
  it('reads output_text, never source_text', () => {
    expect(
      rowWordCount({ output_text: 'hello world', entry_type: 'transcript' }),
    ).toBe(2);
  });

  it('an image row is always 0 — output_text there is a system descriptor, not speech', () => {
    expect(
      rowWordCount({ output_text: '图片 · PNG · 78 KB', entry_type: 'image' }),
    ).toBe(0);
  });

  it('an empty transcript row is 0 (renders nothing, never "0 字")', () => {
    expect(rowWordCount({ output_text: '', entry_type: 'transcript' })).toBe(0);
  });
});

File: apps/mobile/test/entry_metrics_test.dart
// §4b-8 (docs/strategy/2026-08-01-data-asset-lifecycle-design.md) — pins the
// ONE word-count algorithm and the duration formatter behind the per-row
// 「时长 + 字数」 display. See entry_metrics.dart's header for the judgement
// calls this file locks down (mixed CJK/Latin counting rule; duration
// sub-second branch).

import 'package:flowmic/src/signaling/wire_payloads.dart';
import 'package:flowmic/src/timeline/entry_metrics.dart';
import 'package:flowmic/src/timeline/timeline_entry.dart';
import 'package:flutter_test/flutter_test.dart';

TimelineEntry _entry({
  String source = '',
  String? output,
  String? processMode,
  String? processedText,
  int? durationMs,
  String entryType = TimelineEntry.kTranscript,
}) {
  final DateTime now = DateTime.utc(2026, 8, 1, 10, 0);
  return TimelineEntry(
    id: 'loc_mobile_c',
    clientId: 'c',
    mode: FlowMode.realtime,
    delivery: Delivery.inject,
    sourceText: source,
    outputText: output ?? source,
    processMode: processMode,
    processedText: processedText,
    status: EntryStatus.injected,
    durationMs: durationMs,
    entryType: entryType,
    createdAt: now,
    updatedAt: now,
  );
}

void main() {
  group('textWordCount — the one mixed CJK/Latin algorithm', () {
    test('empty string is zero', () {
      expect(textWordCount(''), 0);
    });

    test('pure CJK counts one per character', () {
      expect(textWordCount('你好世界'), 4);
    });

    test('pure Latin counts one per whitespace-delimited word', () {
      expect(textWordCount('hello world'), 2);
    });

    test('mixed CJK + Latin + digits sums both rules', () {
      // 你(1) 好(1) + world(1) + 123(1) = 4.
      expect(textWordCount('你好 world 123'), 4);
    });

    test('CJK punctuation contributes nothing', () {
      expect(textWordCount('你好，世界！'), 4);
    });

    test('ASCII punctuation does not split or inflate a Latin run', () {
      // "sent." is one word — the period is not counted, and does not start
      // a second word.
      expect(textWordCount('The report has been sent.'), 5);
    });

    test('a run of digits counts as one word, not one per digit', () {
      expect(textWordCount('room 1234'), 2);
    });

    // ── 2026-08-14: the nine-locale rule (entry_metrics.dart header §A–§F) ──
    // Every number below is MEASURED on this machine, not predicted, and the
    // 「was」 column is the pre-2026-08-14 mobile algorithm measured by running
    // a verbatim copy of it — not by reading it.
    //
    // 🔴 THE ONE THAT WAS A SHIPPING DEFECT, not a convention change:
    // `i18n/mobile/ru.json` went out with 「$n слов」 hours before this test
    // existed, so a Russian user read `0 слов` on every row, a live draft
    // counter frozen at 0, and 0 in the statistics page — the permanently-zero
    // cell CLAUDE.md and 16 册 §6.1 forbid by name.
    test('🔴 Cyrillic counts its words — it used to return ZERO', () {
      expect(textWordCount('привет мир'), 2); // was 0
      expect(textWordCount('Проверка'), 1); // was 0
      // The reason the old rule hit zero: NOTHING in the string was admitted,
      // so there was no partial credit to notice. A single Latin word next to
      // it would have made the row look plausible while still undercounting.
      expect(textWordCount('привет world'), 2); // was 1
    });

    test('accented Latin is one word, not one word per accent', () {
      // `ü` used to fall out of the run and split the word in two.
      expect(textWordCount('Grüße aus München'), 3); // was 5
      expect(textWordCount("Ça va, l'été"), 3); // was 4
      expect(textWordCount('Ñandú español'), 2); // was 3
    });

    // 🔴 A CONVENTION CHANGE, not a repair — it changes a number the user sees.
    // Korean is written with spaces between words; that is exactly what
    // separates it from Chinese and Japanese, and the translation industry
    // counts it space-delimited, in 어절 (eojeol). Counting each Hangul
    // syllable is the Korean equivalent of counting English letters.
    test('Korean counts 어절 (space-delimited), not syllables', () {
      expect(textWordCount('한국어 단어 세기'), 3); // was 7
      expect(textWordCount('안녕하세요'), 1); // was 5 — one run, one word
    });

    test('Chinese and Japanese are UNCHANGED — they have no word spaces', () {
      expect(textWordCount('你好世界'), 4);
      expect(textWordCount('日本語のテキスト'), 8);
    });

    test('a run with no letter or digit scores nothing', () {
      // ⚠️ Already true of the old mobile rule (measured: all three were 0
      // before as well). It is pinned here because mobile now runs the DESKTOP
      // algorithm, where this WAS a real change (「！！！」 scored 1 there), and
      // the shared rule needs the case held down on both ends.
      expect(textWordCount('！！！'), 0);
      expect(textWordCount('...'), 0);
      expect(textWordCount('---'), 0);
      // Knowingly accepted, same as the desktop twin: an emoji-only utterance
      // counts 0. 「0」 is the honest answer for a length estimate whose unit is
      // 字/词 — neither is present — and the row still renders its text.
      expect(textWordCount('🙂🙂'), 0);
    });

    test('punctuation ABSORBED into a lettered run still scores that run once',
        () {
      // The run 「，world」 has letters, so it scores 1 — the comma rides along
      // rather than being counted or splitting the word.
      expect(textWordCount('你好，world'), 3);
      expect(textWordCount('你好， world'), 3);
      expect(textWordCount('好!'), 1);
      expect(textWordCount('v0.2.36, 好的!'), 3); // was 5
    });
  });

  group('entryWordCount — reads the MODE\'S FINAL RESULT (owner §4-2)', () {
    test('realtime row (no processMode): counts the plain transcript', () {
      final TimelineEntry e = _entry(source: '你好世界');
      expect(entryWordCount(e), 4);
    });

    // 🔴 THE POINT OF THIS TEST — source and output are deliberately
    // different LENGTHS (12 CJK chars vs. 2 English words) so a function
    // that accidentally reads sourceText instead of the final result cannot
    // pass by coincidence. This is the assertion the reverse-control run
    // (delivery report) breaks on purpose.
    test('translate row: counts the TRANSLATED text, not the 原文', () {
      final TimelineEntry e = _entry(
        source: '这是一段比较长的原始文字内容',
        output: 'Short translation.',
        processMode: 'translate',
        processedText: 'Short translation.',
      );
      expect(entryWordCount(e), 2, reason: '"Short translation." = 2 words');
      expect(entryWordCount(e), isNot(textWordCount(e.sourceText!)));
    });

    test('organize row: counts the ORGANIZED text, not the 原文', () {
      final TimelineEntry e = _entry(
        source: '嗯那个我们今天开会主要是想讨论一下这个事情',
        output: '今天开会讨论此事。',
        processMode: 'organize',
        processedText: '今天开会讨论此事。',
      );
      expect(entryWordCount(e), 8, reason: '今天开会讨论此事 = 8 CJK chars');
      expect(entryWordCount(e), isNot(textWordCount(e.sourceText!)));
    });

    test('a picture row has no transcript — null, not a fabricated zero', () {
      final TimelineEntry e = _entry(
        source: '🖼 PNG · 78 KB',
        entryType: TimelineEntry.kImage,
      );
      expect(entryWordCount(e), isNull);
    });
  });

  group('formatEntryDuration — boundaries', () {
    test('zero', () {
      expect(formatEntryDuration(0), '0.0s');
    });

    test('under one second keeps one decimal', () {
      expect(formatEntryDuration(500), '0.5s');
      expect(formatEntryDuration(999), '1.0s'); // rounds, still the <1s branch
    });

    test('exactly one second and up to 59s are whole seconds', () {
      expect(formatEntryDuration(1000), '1s');
      expect(formatEntryDuration(1999), '1s'); // floors, matches recording
      //  panel's whole-second convention above the sub-second branch
      expect(formatEntryDuration(59999), '59s');
    });

    test('one minute and above switches to m:ss', () {
      expect(formatEntryDuration(60000), '1:00');
      expect(formatEntryDuration(61000), '1:01');
      expect(formatEntryDuration(125000), '2:05');
    });

    test('defensive floor on a negative input (never expected from a caller)', () {
      expect(formatEntryDuration(-500), '0.0s');
    });
  });
}
